federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java

```diff
@@ -510,37 +511,44 @@ private Condition createSearchCondition(LDAPQueryConditionsBuilder conditionsBui
+        // get the attributes that are managed by the configured ldap mappers
+        Set<String> managedAttrs = realm.getComponentsStream(model.getId(), LDAPStorageMapper.class.getName())
+                .map(mapperManager::getMapper)
+                .map(LDAPStorageMapper::getUserAttributes)
+                .flatMap(Set::stream)
+                .collect(Collectors.toSet());
 
         try (LDAPQuery ldapQuery = LDAPUtils.createQueryForUserSearch(this, realm)) {
 
             LDAPQueryConditionsBuilder conditionsBuilder = new LDAPQueryConditionsBuilder();
 
-            // Mapper should replace parameter with correct LDAP mapped attributes
-            if (attributes.containsKey(UserModel.USERNAME)) {
-                ldapQuery.addWhereCondition(conditionsBuilder.equal(UserModel.USERNAME, attributes.get(UserModel.USERNAME)));
-            }
-            if (attributes.containsKey(UserModel.EMAIL)) {
-                ldapQuery.addWhereCondition(conditionsBuilder.equal(UserModel.EMAIL, attributes.get(UserModel.EMAIL)));
-            }
-            if (attributes.containsKey(UserModel.FIRST_NAME)) {
-                ldapQuery.addWhereCondition(conditionsBuilder.equal(UserModel.FIRST_NAME, attributes.get(UserModel.FIRST_NAME)));
-            }
-            if (attributes.containsKey(UserModel.LAST_NAME)) {
-                ldapQuery.addWhereCondition(conditionsBuilder.equal(UserModel.LAST_NAME, attributes.get(UserModel.LAST_NAME)));
+            for (Map.Entry<String, String> entry : attributes.entrySet()) {
+                String attrName = entry.getKey();
+                if (LDAPConstants.LDAP_ID.equals(attrName)) {
+                    String uuidLDAPAttributeName = this.ldapIdentityStore.getConfig().getUuidLDAPAttributeName();
+                    Condition usernameCondition = conditionsBuilder.equal(uuidLDAPAttributeName, entry.getValue());
+                    ldapQuery.addWhereCondition(usernameCondition);
+                } else if (LDAPConstants.LDAP_ENTRY_DN.equals(attrName)) {
+                    ldapQuery.setSearchDn(entry.getValue());
+                    ldapQuery.setSearchScope(SearchControls.OBJECT_SCOPE);
+                } else if (managedAttrs.contains(attrName)) {
+                    // we can search any attribute that is mapped to a user attribute
+                    ldapQuery.addWhereCondition(conditionsBuilder.equal(attrName, entry.getValue()));
+                } else if (!attrName.startsWith(UserModel.USER_QUERY_PREFIX)
+                        && !(UserModel.ENABLED.equals(attrName) && Boolean.parseBoolean(entry.getValue()))) {
+                    // if the attr is not mapped just return empty stream
+                    // skip special names and enabled if looking for true (enabled is not mapped so it's always true)
+                    logger.debugf("Searching in LDAP using unmapped attribute [%s], returning empty stream", attrName);
+                    return Stream.empty();
```

Comment:
I am not sure about this last `if`. I can't recall the reason but I think there was some why we wanted to ignore params that we can't search in LDAP. @vramik Can you recall why we ignored them?

Why do attributes with the `USER_QUERY_PREFIX` have different handling? Why these can be ignored during the search and others will result in an empty stream?

It makes sense to me to return an empty stream if we are searching for `ENABLED=false` though.

Reply:
> Why do attributes with the USER_QUERY_PREFIX have different handling? Why these can be ignored during the search and others will result in an empty stream?

There are special attributes like `keycloak.session.realm.users.query.exact` or `keycloak.session.realm.users.query.include_service_accounts` that have specific meanings for the internal DB (doing a exact search without adding prefix search, include service-accounts users,...) and are those parameters are sent to the method. Those attributes are not taken into account in the LDAP.

Regarding the last if, I was returning the empty stream because if you search the ldap via an unknown attribute to the ldap I think we should return nothing. For example, if you search users for a realm with two ldap stores, and there is an attribute managed in one ldap but not in the other. My understanding is that the first ldap should return the users with the passed attr value, but no user from the second ldap (no user has that attribute). Returning all users for the second is counterintuitive, at least for me. :smile:

Reply:
> There are special attributes like keycloak.session.realm.users.query.exact or keycloak.session.realm.users.query.include_service_accounts that have specific meanings for the internal DB (doing a exact search without adding prefix search, include service-accounts users,...) and are those parameters are sent to the method. Those attributes are not taken into account in the LDAP.

But if we ignore these it means we may return users that do not fulfill the searching criteria. Like if we search users with `IDP_ALIAS` shouldn't we return empty stream from LDAP too? 

> Regarding the last if, I was returning the empty stream because if you search the ldap via an unknown attribute to the ldap I think we should return nothing. For example, if you search users for a realm with two ldap stores, and there is an attribute managed in one ldap but not in the other. My understanding is that the first ldap should return the users with the passed attr value, but no user from the second ldap (no user has that attribute). Returning all users for the second is counterintuitive, at least for me. 😄

I agree with you, it makes sense to me to return empty stream but I am not sure this won't break something. Let's wait for a response from @vramik if he recalls some reason we had it this way. If we won't find any reason we can proceed as it makes sense what you described.

Reply:
> But if we ignore these it means we may return users that do not fulfill the searching criteria. Like if we search users with IDP_ALIAS shouldn't we return empty stream from LDAP too?

Yep, you are right, we need to discard names individually (INCLUDE_SERVICE_ACCOUNT and EXACT only). I'll modify the PR when @vramik comments about returning the empty stream or not.

Reply:
@vramik Remember to take a look to this when you have time. :smile:

Reply:
@rmartinc @mhajas sorry for late reply. 

> I can't recall the reason but I think there was some why we wanted to ignore params that we can't search in LDAP. @vramik Can you recall why we ignored them?

There are probably historical reasons why only these are supported. When you check the changes before, it looked: https://github.com/keycloak/keycloak/blame/b6a4b0f80349713dbc378dd346e87bd32ae10726/federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java#L367-L393 and if I go even further into history it seems only those params were supported since the beginning: https://github.com/keycloak/keycloak/commit/c75dcb90c2bb32e2ced12fccc7d6bf2ddf14cda0#diff-ba4d561418b68c20b6081e1cb95c9794f9817f42671358d6d8abbe087d3a7184R237-R254

Unless the proposed approach has some e.g. performance drawbacks I think it's good to go ahead with it.